Mathlib/LinearAlgebra/Matrix/DFT.lean

```diff
@@ -0,0 +1,274 @@
+  apply invertibleOfRightInverse  _ (of (fun (k p : Fin n) => exp (-2 * π * I * k * p / n) / n)) _
+  funext k p
+  -- rw [mul_apply]
+  simp_rw [mul_apply, Wₙ_apply, of_apply, mul_div_assoc', ← Complex.exp_add]
+  by_cases h : k = p
+  · rw [h, one_apply_eq]
+    ring_nf
+    simp_rw [Complex.exp_zero, Finset.sum_const, card_fin, nsmul_eq_mul]
+    apply mul_inv_cancel_left₀ (Nat.cast_ne_zero.2 (NeZero.ne _))
+  · rw [one_apply_ne h]
+    have h1 : ∀ (x : Fin n), 2 * π * I * k * x / n +  (-2) * π * I * x * p / n  =
+        (x : ℕ) * ((2*π * I * (k - p)) / n) := by intro x; ring
+    simp_rw [h1, ← sum_div, div_eq_zero_iff, Complex.exp_nat_mul, Fin.sum_univ_eq_sum_range]
+    left
+    rw [geom_sum_eq, ← Complex.exp_nat_mul, div_eq_zero_iff]
+    left
+    rw [sub_eq_zero, Complex.exp_eq_one_iff]
+    use (k - p)
+    rw [mul_div_cancel', mul_comm, mul_eq_mul_right_iff]
+    left
+    simp_rw [Int.cast_sub, Int.cast_ofNat]
+    exact (Nat.cast_ne_zero.2 (NeZero.ne _))
+    apply cexp_sub_ne_one _ _ h
+
+/-- The IDFT Matrix -/
+theorem dftMatrix_inv  [Invertible (Wₙ n)] :
+    ⅟(Wₙ n) = of (fun (k p : Fin n) => exp ( -2 * π * I * k * p / n) / n) := by
+  letI := (invWₙ n)
+  convert (rfl : ⅟(Wₙ n) = _)
+
+lemma iWₙ_apply (k p : Fin n) : (Wₙ n)⁻¹ k p = exp (-2 * π * I * k * p / n)/n := by
+  letI := invWₙ n
+  rw [← Matrix.invOf_eq_nonsing_inv (Wₙ n), dftMatrix_inv, of_apply]
+
+/-- The DFT operation and DFT matrix applied to sequence/vector are tha same -/
+theorem dft_eq_Wₙ_mul (v : Fin n → ℂ) : dft n v = mulVec (Wₙ n) v := by
+  funext r
+  simp only [dft, mulVec, dotProduct, Wₙ_apply]
+
+/-- The IDFT operation and IDFT matrix applied to sequence/vector are tha same -/
+theorem idft_eq_iWₙ_mul (V : Fin n → ℂ ) : idft n V = mulVec (Wₙ n)⁻¹ V := by
+  funext r
+  simp only [idft, mulVec, dotProduct, iWₙ_apply]
+
+/-- IDFT/DFT are bijective pair-/
+theorem idft_dft  (v : Fin n → ℂ) : idft n (dft n v) = v := by
+  letI := invWₙ n
+  rw [dft_eq_Wₙ_mul, idft_eq_iWₙ_mul, mulVec_mulVec, inv_mul_of_invertible, one_mulVec]
+
+theorem dft_idft  (V : Fin n → ℂ) : dft  n (idft n V) = V := by
+  letI := invWₙ n
+  rw [dft_eq_Wₙ_mul, idft_eq_iWₙ_mul, mulVec_mulVec, mul_inv_of_invertible, one_mulVec]
+
+/-- The IDFT matrix is the scaled conjugate transpose of the DFT matrix-/
+theorem Wₙ_conjTranspose_eq_iWₙ :  (Wₙ n)⁻¹ = ((1:ℂ)/n) • (Wₙ n)ᴴ := by
+  funext x y
+  simp only [iWₙ_apply, smul_apply, conjTranspose_apply, Wₙ_apply, smul_eq_mul]
+  rw [star_def, ← Complex.exp_conj, ← div_mul_comm, mul_one, div_left_inj', ← star_def]
+  simp only [star_div', star_natCast, star_mul', conj_I, star_def, conj_ofReal, map_ofNat, neg_mul,
+    mul_neg]
+  ring_nf
+  exact (Nat.cast_ne_zero.2 (NeZero.ne _))
+
+/-- The DFT matrix is symmetric -/
+lemma Wₙ_transpose_eq_Wₙ : (Wₙ n)ᵀ = Wₙ n := by
+  funext a b
+  simp only [transpose_apply, Wₙ_apply]
+  ring_nf
+
+def shiftk (N : ℕ) (k : Fin N) : (Fin N → Fin N) := fun n : (Fin N) => (n - k)
```

Comment:
I'm fairly sure this exists somewhere in https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html